Workaround #28126, support SIMDing broadcast in more cases

mbauman opened the pull request:
This is an ugly performance hack around issue #28126 in some limited (but common) cases. The problem in short: when given many arrays of the same size, LLVM has difficulty hoisting the decision of whether a given dimension should be "extruded" out of the loop. This extra indirection in the index computation seems to foil the array bounds aliasing checks, which stymies SIMDification. The solution: check to see if _Julia_ can statically decide whether or not to extrude any dimensions in a given broadcast expression -- and if so, use a special array wrapper that flags that none of the dimensions in that array need to be extruded out in order to perform the broadcast.

No idea if tests will pass or if this actually is a good idea in general.

@nanosoldier `runbenchmarks(ALL, vs = ":master")` (edit: whoops, sorry Nanosoldier — should have waited for CI before sending you down a dead-end alley)

mbauman committed:
Workaround #28126, support SIMDing broadcast in more cases

This is an ugly performance hack around issue #28126 in some limited (but common) cases. The problem in short: when given many arrays of the same size, LLVM has difficulty hoisting the decision of whether a given dimension should be "extruded" out of the loop. This extra indirection in the index computation seems to foil the array bounds aliasing checks, which stymies SIMDification. The solution: check to see if _Julia_ can statically decide whether or not to extrude any dimensions in a given broadcast expression -- and if so, use a special array wrapper that flags that none of the dimensions in that array need to be extruded out in order to perform the broadcast.

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/30973):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`make -j3 USECCACHE=1`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/fcfc0f444278466f3338535b84dcaeba193b1028/16231c4_vs_9562bdf/report.md)
cc @ararslan

mbauman committed:
Fixup NonExtruded indexing

YingboMa commented on base/broadcast.jl:
```diff
@@ -864,13 +864,14 @@ broadcast_unalias(::Nothing, src) = src
-preprocess_args(dest, args::Tuple{Any}) = (preprocess(dest, args[1]),)
-preprocess_args(dest, args::Tuple{}) = ()
+@inline preprocess_args(f, dest, args::Tuple) = (preprocess(f, dest, args[1]), preprocess_args(f, dest, tail(args))...)
+preprocess_args(f, dest, args::Tuple{Any}) = (preprocess(f, dest, args[1]),)
```
Changing this to
```julia
Base.@propagate_inbounds preprocess_args(f, dest, args::Tuple) = (preprocess(f, dest, args[1]), preprocess_args(f, dest, tail(args))...)
Base.@propagate_inbounds preprocess_args(f, dest, args::Tuple{Any}) = (preprocess(f, dest, args[1]),)
```
get rid of the memory allocation in
```julia
using BenchmarkTools
u, k1, k2, k3, k4, k5, k6, k7 = (ones(1000) for i in 1:8)
function goo(u, k1, k2, k3, k4, k5, k6, k7)
  @. u = 0.1*(0.1*k1 + 0.2*k2 + 0.3*k3 + 0.4*k4 + 0.5*k5 + 0.6*k6 + 0.7*k7)
  nothing
end
@btime goo($u, $k1, $k2, $k3, $k4, $k5, $k6, $k7)
```
Before:
```julia
julia> @btime goo($u, $k1, $k2, $k3, $k4, $k5, $k6, $k7)
  707.291 ns (16 allocations: 416 bytes)
```
After:
```julia
julia> @btime goo($u, $k1, $k2, $k3, $k4, $k5, $k6, $k7)
  640.183 ns (0 allocations: 0 bytes)
```

mbauman replied:
Thanks — the key is just `@inline`.

mbauman committed:
inline and add test case

mbauman commented:
Ok, let's try this again. @nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/30973#issuecomment-461079574) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8ef4827edfe17556425c33e1e951cdbab4cb48d1/e5c4492_vs_f97c443/report.md). cc @ararslan

YingboMa commented:
Is it possible to remove all the allocation in here?
```julia
tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13,
     k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25,
     k26, k27, k28, k29, k30, k31, k32, k33, k34 = (ones(1000) for i in 1:36)

function foo(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13,
             k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25,
             k26, k27, k28, k29, k30, k31, k32, k33, k34)
  dt = 0.5
  @. tmp = uprev + dt*(0.00*k1 + 0.01*k2 + 0.02*k3 + 0.04*k5 + 0.06*k7 + 0.07*k8 + 0.09*k10 + 0.10*k11 + 0.11*k12 + 0.12*k13 + 0.13*k14 + 0.14*k15 + 0.15*k16 + 0.16*k17 + 0.17*k18 + 0.18*k19 + 0.19*k20 + 0.20*k21 + 0.21*k22 + 0.22*k23 + 0.23*k24 + 0.24*k25 + 0.25*k26 + 0.26*k27 + 0.27*k28 + 0.28*k29 + 0.29*k30 + 0.30*k31 + 0.31*k32 + 0.32*k33 + 0.33*k34)
  nothing
end

foo(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k30, k31, k32, k33, k34)
@allocated foo(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13,
             k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25,
             k26, k27, k28, k29, k30, k31, k32, k33, k34)
```
```julia
julia> @allocated foo(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13,
                    k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25,
                    k26, k27, k28, k29, k30, k31, k32, k33, k34)
7072
```
This isn't some arbitrary example. It is a line in the `Feagin14` method. https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/perform_step/feagin_rk_perform_step.jl#L737

YingboMa committed:
Inline the whole broadcast expression to avoid allocation (#30986)

* Inline the whole broadcast expression to avoid allocation

`map` has an inline limit of 16. To make sure that the whole broadcast
tree gets inlined properly, I added the `_inlined_map` function. I am not
sure if it is a good idea, but worth trying.

This PR solves the issue which I have mentioned in
https://github.com/JuliaLang/julia/pull/30973/commits/2693778d0fc8696280b540e9135ea0220cf6d6a7#issuecomment-461248258

```julia
julia> @allocated foo(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13,
                    k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25,
                    k26, k27, k28, k29, k30, k31, k32, k33, k34)
0
```

* Fix CI failure

* Stricter test (vectorization & no allocation for a 9-array bc)

* rm `_inlined_map`

KristofferC commented:
Perhaps some compilation time comparisons before after this PR would be useful to see if there is a trade off.

mbauman committed:
Move vectorization test to boundscheck

mbauman commented:
So I tried a bunch of compile time tests (Julia build time, using JuliaDB/DataFrames/Plots, `plot(rand(10))`, various Julia tests), but the only one that showed any change was test/broadcast:

### master (9f69b2e4c9)

```
$ time make -C test broadcast
make: Entering directory '/home/mbauman/julia/test'
    JULIA test/broadcast
Test  (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
running testset broadcast...
broadcast  (1) |    44.72 |   2.39 |  5.3 |    5557.00 |   344.13

Test Summary: | Pass  Total
  Overall     |  423    423
    SUCCESS
make: Leaving directory '/home/mbauman/julia/test'

real	0m49.138s
user	0m48.738s
sys	0m1.190s
```

## mb/hack28126 (8aa59077da)

```
$ time make -C test broadcast
make: Entering directory '/home/mbauman/julia/test'
    JULIA test/broadcast
Test  (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
running testset broadcast...
broadcast  (1) |    53.04 |   2.99 |  5.6 |    7046.15 |   370.40

Test Summary: | Pass  Total
  Overall     |  424    424
    SUCCESS
make: Leaving directory '/home/mbauman/julia/test'

real	0m57.454s
user	0m56.996s
sys	0m1.190s
```

That's about a 20% regression.  :-\

chriselrod commented:
Is this being discussed on triage?

mbauman commented:
The PR to revive (first) is #35675. Then the concern in https://github.com/JuliaLang/julia/pull/30973#discussion_r254351339 vanishes.

chriselrod commented:
> Not sure what to do here; is it worth trying to revive this? Does the problem still exist?

Yes.
```julia
using FastBroadcast, BenchmarkTools
const tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13,
     k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25,
     k26, k27, k28, k29, k30, k31, k32, k33, k34 = (ones(1000) for i in 1:36)

function foo(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13,
             k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25,
             k26, k27, k28, k29, k30, k31, k32, k33, k34)
  dt = 0.5
  @. tmp = uprev + dt*(0.00*k1 + 0.01*k2 + 0.02*k3 + 0.04*k5 + 0.06*k7 + 0.07*k8 + 0.09*k10 + 0.10*k11 + 0.11*k12 + 0.12*k13 + 0.13*k14 + 0.14*k15 + 0.15*k16 + 0.16*k17 + 0.17*k18 + 0.18*k19 + 0.19*k20 + 0.20*k21 + 0.21*k22 + 0.22*k23 + 0.23*k24 + 0.24*k25 + 0.25*k26 + 0.26*k27 + 0.27*k28 + 0.28*k29 + 0.29*k30 + 0.30*k31 + 0.31*k32 + 0.32*k33 + 0.33*k34)
  nothing
end
function foo_fb(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13,
             k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25,
             k26, k27, k28, k29, k30, k31, k32, k33, k34)
  dt = 0.5
  @.. tmp = uprev + dt*(0.00*k1 + 0.01*k2 + 0.02*k3 + 0.04*k5 + 0.06*k7 + 0.07*k8 + 0.09*k10 + 0.10*k11 + 0.11*k12 + 0.12*k13 + 0.13*k14 + 0.14*k15 + 0.15*k16 + 0.16*k17 + 0.17*k18 + 0.18*k19 + 0.19*k20 + 0.20*k21 + 0.21*k22 + 0.22*k23 + 0.23*k24 + 0.24*k25 + 0.25*k26 + 0.26*k27 + 0.27*k28 + 0.28*k29 + 0.29*k30 + 0.30*k31 + 0.31*k32 + 0.32*k33 + 0.33*k34)
  nothing
end

@benchmark foo(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k30, k31, k32, k33, k34)
@benchmark foo_fb(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k30, k31, k32, k33, k34)
```
I get:
```julia
julia> @benchmark foo(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k30, k31, k32, k33, k34)
BechmarkTools.Trial: 10000 samples with 1 evaluations.
 Range (min … max):  60.687 μs …  2.011 ms  ┊ GC (min … max): 0.00% … 93.25%
 Time  (median):     61.530 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   63.451 μs ± 41.939 μs  ┊ GC (mean ± σ):  1.56% ±  2.29%

  ▁▅▆██▇▆▅▃▂▂▂▂▃▃▂▂▁                ▁ ▁▁▂▁▁▁▁▁   ▁            ▂
  ██████████████████████▆▇▆▆▆▇▆▇▇▇██████████████████▇▇▇▇▆▅▆▄▆ █
  60.7 μs      Histogram: log(frequency) by time      70.8 μs <

 Memory estimate: 29.80 KiB, allocs estimate: 625.

julia> @benchmark foo_fb(tmp, uprev, k1, k2, k3, k4, k5, k6, k7, k8, k9, k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k30, k31, k32, k33, k34)
BechmarkTools.Trial: 10000 samples with 1 evaluations.
 Range (min … max):  25.045 μs …  2.118 ms  ┊ GC (min … max): 0.00% … 96.72%
 Time  (median):     25.929 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   28.123 μs ± 50.510 μs  ┊ GC (mean ± σ):  4.66% ±  2.56%

  ▁▅▇██▇▆▅▄▃▂▁     ▁▁▁▁ ▁          ▁ ▁ ▁▁▁▁▁▁                 ▂
  ██████████████▇██████████▇▆▇▆▆▇███▇█████████████▇██▇▇▇▆▆▆▆▆ █
  25 μs        Histogram: log(frequency) by time      35.8 μs <

 Memory estimate: 31.94 KiB, allocs estimate: 663.
```

N5N3 commented:
I think there's no need to limit this optimization to inputs with the same axes.
Something like `ExtrudedWithoutDim1` should be enough, as `@simd` only tries to vectorlize the innermost loop.

vtjnash commented:
Seems no longer needed
